docs/source/user/serialization.rst

```diff
@@ -79,6 +79,65 @@ schema:
 For a more complete example of Uplink's :mod:`marshmallow` support,
 check out `this example on GitHub <https://github.com/prkumar/uplink/tree/master/examples/marshmallow>`_.
 
+.. _using_pydantic_schemas:
+
+Using Pydantic Models
+=========================
+
+:mod:`pydantic` is a framework-agnostic, object serialization library
+for Python >= 3.6. Uplink comes with built-in support for Pydantic; you can
+integrate your Pydantic models with Uplink for easy JSON (de)serialization.
+
+First, create a :class:`pydantic.BaseModel`, declaring any necessary
+conversions and validations. Here's a simple example:
+
+.. code-block:: python
+
+   from typing import List
+
+   from pydantic import BaseModel, HttpUrl
+
+   class Owner(BaseModel):
+      id: int
+      avatar_url: HttpUrl
+      organizations_url: HttpUrl
+
+   class Repo(BaseModel):
+      id: int
+      full_name: str
+      owner: Owner
+
+Then, specify the schema using the :class:`@returns <uplink.returns>` decorator:
+
+.. code-block:: python
+   :emphasize-lines: 2
+
+   class GitHub(Consumer):
+      @returns(List[Repo])
+      @get("users/{username}/repos")
+      def get_repos(self, username):
+         """Get the user's public repositories."""
+
+Python 3 users can use a return type hint instead:
+
+.. code-block:: python
+   :emphasize-lines: 3
+
+   class GitHub(Consumer):
+      @returns.json()
+      @get("users/{username}/repos")
+      def get_repos(self, username) -> List[Repo]:
+         """Get the user's public repositories."""
```

Comment:
Hey @prkumar . There's a little implementation gotcha here. I wasn't able to get rid of `returns.json` decorator. IMO pydantic should work as marshmallow, deserializing from json by default.

If I implement some "list parsing" logic [here](https://github.com/prkumar/uplink/pull/200/files#diff-6fbe4db3d7d3ca8d9b12a5a0968eb351R23)
```python
if isinstance(list, data):
  return [self._model(**entry) for entry in data)
```
you get a response of `[[Model1, Model2]]`, thus breaking the interface.
Do you have any idea for such case ?

As a last resort, it's possible to delegate this logic to the user.
```python
class MyModel(BaseModel):
    id: str

# case you expect a list, you must implement a new model.
class ListResponseModel(BaseModel):
    ___root___ = List[MyModel]
```

Reply:
IIUC, this problem would occur with `marshmallow` as well, right? I haven't tried this myself, but my expectation is that something like `List[RepoSchema]`, where `RepoSchema` is a subclass of `marshmallow.Schema`, would require the `@returns.json()` decorator, similarly because of the `List` annotation. In this case, we would have to use `RepoSchema(many=True)` if we wanted to omit the `@returns.json`. 

IMO, since `pydantic` does have a similar options as `many=True`, I  think we should recommend setting `@returns.json()` when using `BaseModel` subclasses, instead of defining a separate `ListResponseModel`. Moreover, it looks like `pydantic` may support formats other than JSON; e.g., the documentation boasts [efficient picking and unpickling](https://pydantic-docs.helpmanual.io/usage/exporting_models/#pickledumpsmodel) of data (although I doubt sending pickled pydantic models over the wire is a secure and widely used practice).

Also, speaking personally, I have grown to prefer library's that allow/require the user to explicitly define the response's expected encoding (e.g., `json` vs `protobuf`) rather than deriving it. Looking back, I kind of wish I had this preference in mind when designing the (de)serialization support for `uplink`: e.g., grouped all mime types under a `content_type` or `decode` module and required the user to explicitly set the expected encoding of the response when using a return annotation to deserialize the response.

Reply:
I totally agree ! The separation of concerns of what(mime)/how(model) would be great ! Using your own example, case a user wanted to send/read a pickle using the same pydantic's model woudl require us to write a new converter. It's a good looking feature from a "dev point of view" but I'm not quite sure about use cases (thinking about a refactor..) 🤔